### PR TITLE
[#1378] Support datetime format with timezone offset

### PR DESCRIPTION
**Pull request by dejanb**
This should allow encoding datetime values both with and without timezone offset. The value will be decoded without the offset.

### CODE REVIEW

File: service-base/pom.xml [hunk 1]
```diff
@@ -43,6 +43,11 @@
       <groupId>org.eclipse.hono</groupId>
       <artifactId>hono-core</artifactId>
     </dependency>
+    <dependency>
+        <groupId>com.fasterxml.jackson.datatype</groupId>
+        <artifactId>jackson-datatype-jsr310</artifactId>
```
Comment (sophokles73) [inline service-base/pom.xml]: we'll need a CQ for this

Comment (dejanb) [reply service-base/pom.xml]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20419

**Comment by sophokles73** [comment]
@dejanb so, in order to be able to at least do the example deployment, we will need to change the date/time format used in the example credentials, right?

**Comment by dejanb** [comment]
@sophokles73 no, this should work properly with the existing credentials. There's a test in this pr to guard that

**Comment by sophokles73** [comment]
@dejanb according to #1355 the credentials cannot be retrieved from the registry, though ...

**Comment by dejanb** [comment]
@sophokles73 this is the pr to fix that :)

**Comment by sophokles73** [comment]
I know, but because we'll have to wait for the CQ to be approved, we cannot merge it (yet). For the time being, though, it would be nice to have a working system, don't you think?

**Comment by dejanb** [comment]
Depends on how long it takes for this to be merged. I would like to have a release with the original example credentials for backward compatibility. 
I'll create a pr with the temporal change in the examples.

**Comment by ctron** [comment]
A quick question, and please forgive me that I didn't follow this discussion: but why do we need a timezone? Other than to read old (legacy) JSON files?

**Comment by dejanb** [comment]
@ctron we had this discussion on the chat the other day. Resolution: to make it easier for users to define proper time without the constant need to recalculate to UTC

**Comment by ctron** [comment]
But that would be a feature for a UI, not the storage backend. Because now you need to always recalculate to UTC when checking. Which might be more frequent than editing.

**Comment by dejanb** [comment]
This is just one more option for the user, considering that we don't have an UI :)

**Comment by ctron** [comment]
Can we suggest to prefer UTC.

File: service-base/pom.xml (hunk 1 above)
Comment (Alfusainey) [reply service-base/pom.xml]: any news on this yet?

File: service-base/src/test/java/org/eclipse/hono/service/management/credentials/SecretsTest.java [hunk 2]
```diff
@@ -134,4 +143,30 @@ public void testDecodeGeneric() {
...
+                .put(FIELD_SECRETS_NOT_BEFORE, "2017-05-01T14:00:00+01:00")
+                .put(FIELD_SECRETS_NOT_AFTER, "2037-06-01T14:00:00Z");
+
+
+        final PasswordSecret secret = json.mapTo(PasswordSecret.class);
+        final LocalDateTime before = LocalDateTime.of(2017, 05, 01, 14, 00, 00);
+        assertEquals(before.toInstant(ZoneOffset.of("+01:00")), secret.getNotBefore());
+
+        final JsonObject decodedSecret = JsonObject.mapFrom(secret);
+        assertEquals("2017-05-01T13:00:00Z", decodedSecret.getValue("not-before"));
```
Comment (Alfusainey) [inline service-base/src/test/java/org/eclipse/hono/service/management/credentials/SecretsTest.java]: @dejanb Don't you think the expected value in this case should be: `2017-05-01T14:00:00+01:00`, exactly as was specified in the JSON object and not the value in `DateTimeFormatter.ISO_INSTANT` format?

Comment (dejanb) [reply service-base/src/test/java/org/eclipse/hono/service/management/credentials/SecretsTest.java]: @Alfusainey the idea is that ISO_INSTANT is preferred format and that we just supporting defining dates in different formats for better usability.

**Comment by dejanb** [comment]
@sophokles73 CQ has been approved. Do you want to proceed with this or keep the manual formatting?

**Comment by sophokles73** [comment]
> Do you want to proceed with this or keep the manual formatting?

At the moment I do not see much benefit in introducing the additional dependency. Do you?

File: service-base/src/test/java/org/eclipse/hono/service/management/credentials/SecretsTest.java (hunk 2 above)
Comment (Alfusainey) [reply service-base/src/test/java/org/eclipse/hono/service/management/credentials/SecretsTest.java]: thanks 👍

**Comment by sophokles73** [comment]
@dejanb with #1403 around the corner it probably makes sense to rebase and merge this PR. WDYT?

File: service-base/pom.xml (hunk 1 above)
Comment (sophokles73) [reply service-base/pom.xml]: the CQ is approved

File: service-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonSecret.java [hunk 3]
```diff
@@ -52,116 +49,20 @@ public void setEnabled(final Boolean enabled) {
...
-     * Gets the earliest instant in time that this secret may be used for authenticating a device.
-     * 
-     * @return The instant.
-     */
-    @JsonIgnore
     public Instant getNotBefore() {
-        return Optional.ofNullable(notBefore)
-                .map(nb -> nb.toInstant())
-                .orElse(null);
+        return notBefore;
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonSecret.java]: can we agree to keep the JavaDoc?

File: service-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonSecret.java [hunk 4]
```diff
@@ -52,116 +49,20 @@ public void setEnabled(final Boolean enabled) {
...
-     * @param notBefore The instant.
-     */
-    @JsonIgnore
     public void setNotBefore(final Instant notBefore) {
-        if (notBefore == null) {
-            this.notBefore = null;
-        } else {
-            this.notBefore = OffsetDateTime.ofInstant(notBefore, ZoneId.systemDefault());;
-        }
+        this.notBefore = notBefore;
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonSecret.java]: keep JavaDoc?

File: service-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonSecret.java [hunk 5]
```diff
@@ -52,116 +49,20 @@ public void setEnabled(final Boolean enabled) {
...
-     * @see DateTimeFormatter#ISO_OFFSET_DATE_TIME
-     */
-    @JsonSetter(FIELD_SECRETS_NOT_AFTER)
-    public void setNotAfter(final String notAfter) {
-        try {
-            this.notAfter = DateTimeFormatter.ISO_OFFSET_DATE_TIME.parse(notAfter, OffsetDateTime::from);
-        } catch (DateTimeParseException e) {
-            this.notAfter = null;
-        }
+        return notAfter;
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonSecret.java]: keep JavaDoc?

File: service-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonSecret.java [hunk 6]
```diff
@@ -52,116 +49,20 @@ public void setEnabled(final Boolean enabled) {
...
-     * @param notAfter The instant.
-     */
-    @JsonIgnore
     public void setNotAfter(final Instant notAfter) {
-        if (notAfter == null) {
-            this.notAfter = null;
-        } else {
-        this.notAfter = OffsetDateTime.ofInstant(notAfter, ZoneId.systemDefault());
-        }
+        this.notAfter = notAfter;
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonSecret.java]: keep JavaDoc?

**Comment by dejanb** [comment]
@sophokles73 this should be now good to go. Can you give it another look?

**Pull request merged**